File: golang-version/KunalSheet/Strings/longest_common_prefix.go
package main

import (
	"fmt"
	"math"
	"strings"
)

// input: strs = ["flower", "flow", "flight"], output : "flu"

func main() {
	strs := []string{"flower", "flow", "flight"}
	fmt.Println(longestCommonPrefix(strs))
}

func longestCommonPrefix(strs []string) string {
	if strs == nil || len(strs) == 0 {
		return ""
	}

	minLen := math.MaxInt32
	for _, str := range strs {
		if len(str) < minLen {
			minLen = len(str)
		}
	}

	low, high := 1, minLen

	for low <= high {
		middle := (low + high) / 2
		if isCommonPrefix(strs, middle) {
			low = middle + 1
		} else {
			high = middle - 1
		}
	}
	return strs[0][:(low+high)/2]
}

func isCommonPrefix(strs []string, length int) bool {
	prefix := strs[0][:length]
	for i := 1; i < len(strs); i++ {
		if !strings.HasPrefix(strs[i], prefix) {
			return false
		}
	}
	return true
}

File: KunalSheet/src/main/java/org/ds/string/easy/longestCommonPrefix.java
package org.ds.string.easy;

public class longestCommonPrefix {
    public static void main(String[] args){
        
    }

    public static String longestCommonPrefix(String[] strs){
        if (strs == null || strs.length == 0) return "";
        int minLen = Integer.MAX_VALUE;
        for(String str: strs) {
            minLen = Math.min(str.length(), minLen);
        }
        int low = 1;
        int high = minLen;
        while (low <= high){
            int middle= (low + high)/2;
            if(isCommonPrefix(strs,middle)){
                low = middle + 1;
            } else {
                high = middle - 1;
            }
        }
        return strs[0].substring(0, (low+high)/2);
    }

    private static boolean isCommonPrefix(String[] strs, int middle){
        String str1 = strs[0].substring(0,middle);
        for (int i = 1; i< strs.length; i++){
            if(!strs[i].startsWith(str1)){
                return false;
            }
        }
        return true;
    }
}
